### PR TITLE
Add parametric softplus non-linearity

### PR DESCRIPTION
**Pull request by the-moliver**
This is meant to complement this PR: https://github.com/fchollet/keras/pull/421
This non-linearity is useful for modeling neural activity, especially when combined with Poisson loss.

With these additions it is possible to implement all of the models from this paper in keras:
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003143

### CODE REVIEW

**Comment by fchollet** [comment]
Looking good. I think "Psoftplus" is not a very good name, how about ParametricSoftplus? 

Otherwise, this would need documentation. Also, I'm thinking we should probably turn the advanced activations into MaskedLayers, since they should be compatible with mask passing.

**Comment by iskandr** [comment]
@fchollet Do you have a plan for which layers will and won't support masking? I seem to need masked 1D convolutions (https://github.com/fchollet/keras/issues/411) but I'm wondering if in the course of playing around with sequence predictors I'll end up needing every layer to support masks.

**Comment by fchollet** [comment]
> Do you have a plan for which layers will and won't support masking?

What will support masking: the upcoming TimeDistributed layer. All RNN layers or layers dealing with a time dimension. All layers that are shape-agnostic (like these non-linearities).

What won't support (time) masking: all layers that have a defined input shape of which time is not a dimension.

**Comment by the-moliver** [comment]
Updated with requested changes. Let me know if there's anything else.

**Comment by linxihui** [comment]
Actually I have a question:

According to the documentation, _parametric-softmax_ is `alpha * log(1 + exp(beta * x))`, where `alpha` and `beta` are trainable.

But since `x` comes from a linear transform with trainable weights from the previous layer, I feel that a trainable `beta` is not necessary, as it can be merged to the trainable weights.  Similarly, the result of _parametric-softmax_ will be passed to next layer where another trainable linear transform happens (I think _parametric-softmax_ will not used as final output), therefore `alpha` can be merged to weights in next layer.  

I mean, if the previous layer and next layer has trainable weights, the activation with trainable `alpha`, `beta` should yield **exactly** the same result as when they are fixed to 1, mathematically. Numerically, it probably makes optimization more difficult. 

The only case I see a difference is when the previous or next layer has **non-linear** transform or **fixed** weights.

What do you think? 

Eric

**Comment by the-moliver** [comment]
Well technically that is true, but it's really an issue is scale. The parametric softplus is basically a soft rectifier with the softness controlled by alpha and beta. This is easier to see if alpha =1/beta, which is how they are initialized. As beta becomes large and alpha is the inverse, it approaches ReLU. Changing a set of weights of the input layer to be 5 times larger, equivalent to beta =5, would require a lot of learning to just accomplish making the rectifier sharper.
